### ISSUE TITLE
Easy conversion between Result

### ISSUE THREAD

**Issue by Yamakaky** [rust-lang/rfcs#1813]
```rust
impl From<Result<T, E1>> for Result<T, E2> where E2: From<E1>
```

**Comment by burdges** [comment]
I think `impl<T,E1,E2> From<Result<T, E1>> for Result<T, E2> where E2: From<E1>` conflicts with the reflexive `impl<T> From<T> for T` because you cannot constrain it to `E1 != E2`.

**Comment by Yamakaky** [comment]
We need trait specialisation badly...

**Comment by bluss** [comment]
Isn't this the domain of the `?` operator? Together with the catch block, this can be done inline, something like `catch { x? }` (unsure about exact syntax since catch doesn't exist yet).

**Comment by Yamakaky** [comment]
It doesn't work for things like
```rust
fn func() -> Result<()> {
    Err("error")
}
```
I have to use `map_err`.

**Comment by sfackler** [comment]
`Err("error".into())`?

**Comment by Yamakaky** [comment]
Yeah, or `Err("error").into()`. Take in mind that you don't always have an error litteral, if the last call in the function returns a Result.

**Comment by burdges** [comment]
There are many situations where you need a custom result type anyways for various reasons.  In those cases, you could try something like : 
```
pub struct MyResult<T,E>(pub Result<T,E>)
impl<T1,T2,E1,E2> From<MyResult<T1, E1>> for Result<T2, E2> where E2: From<E1>, T2: From<T1>{
    fn from(x: MyResult<T1,E1>) -> Result<T2,E2> {
        match x.0 {
            Ok(y) => Ok(<T2 as From<T1>>::from(y)),
            Err(y) => Err(<E2 as From<E1>>::from(y))
        }
    }
}
```

**Comment by jjyr** [comment]
How is this issue going? No one wants to create an RFC for this feature?

**Comment by burdges** [comment]
As discussed above, there are no viable approaches except for sufficiently powerful specialization, which everyone loves but requires subtle work, and thus moves slowly https://github.com/rust-lang/rust/issues/31844#issuecomment-423389587  You could imagine doing this with more over negative reasoning too, but that's even harder than making specialization work.
